### DISCUSSION TITLE
Brand new Termux API concept?

### DISCUSSION THREAD

**Discussion by twaik** [termux/termux-packages#9100] [Ideas]
Now Termux has termux-api package which helps us with using Android's API's without writing Java code. But it has restrictions:
1. Termux should be in foreground or have `Draw over apps` permission. (https://github.com/termux/TermuxAm/issues/7). Pay attention that it is imposibe to check if we in foreground or check if we have this permission.
2. User should install `Termux::API` application from Google Play or F-Droid.
3. `Termux::API` and `Termux` applications should be signed with the same key.

I propose the following scheme:
0. Make `termux-api` service a part of `termux-app` to fix https://github.com/termux/TermuxAm/issues/7.
1. Use some IPC/RPC with code generation, resource management. That will allow us use API without writing it manually.
2. Start service in separated thread (with simple fork) and keep it to work with unix socket in $PREFIX/var/run as long as Termux's services are alive.

I prefer to use libwayland as IPC because:
0. We can use our IPC as C/C++ library so we can let C/C++ programs to use Android API's (or bindings) without using `termux-api` or other bash scripts.
1. It has resource management so resources created by clients will be alive as long as clients themselves are alive.
2. It has text, non-binary protocol with version control. In the case of protocol version mismatch we can tell the user that his `termux-app` version is outdated.
3. It can send and receive file descriptors so we can i.e. patch libusb to use it. Device enumerating and other stuff will work.
4. There is a [Java scanner](https://github.com/jekstrand/wayland-java/blob/master/scanner/src/main/java/org/freedesktop/wayland/scanner/Scanner.java). We can use this as gradle task which generates protocol implementation in java and use it without writing code in C/C++. This way we do not need to write C bindings to java code.
5. Server side will have access to real `Context` so we can check/ask all the permissions at package installation time (it can be asynchronously, no need to wait while pkg is working) and in runtime (in target program).
6. `termux-gui` also can be reimplemented here...

What do you think about this?

**Comment by twaik** [comment]
@fornwall @Grimler91 @wmcbtech30 @agnostic-apollo

**Reply by twaik** [reply]
I can start writing it if you do not mind.

**Reply by agnostic-apollo** [reply]
The `termux-api` has already moved to using `termux-api-socket` and will be integrated into `termux-app` in coming weeks when I get time to fully test it and make some changes.

https://github.com/termux/termux-am-socket

https://github.com/termux/termux-am-library

https://github.com/termux/termux-api/pull/471

https://github.com/termux/termux-api-package/pull/147

https://github.com/termux/termux-app/pull/2458

`termux-gui` also has a protocol.

https://github.com/termux/termux-gui/blob/main/Protocol.md

**Reply by twaik** [reply]
Good news.

**Reply by twaik** [reply]
But the main points in my idea are making it C/C++ library to make it work inside C programs without any need to use shell, code generation to avoid any mistakes/errors in protocols and resource management which will allow us to create and use Android API objects in Termux process from C/C++ program (and let it be recycled in case if program dies for some reason).

**Reply by twaik** [reply]
I can try to make a demo to show how it works.

**Reply by twaik** [reply]
I mean I can write a demo with a wayland scanner task for Gradle and let Java code create and manage server using generated protocols. But I have a question. Can we embed libwayland and ffi into the termux-app?

**Reply by agnostic-apollo** [reply]
You can make a demo and it can be checked out. It will take time to review though so pre-warning, busy with lot of releases.

> Can we embed libwayland and ffi into the termux-app?

Can you provide links to both. And add in demo or provide info in exactly what will need to be added. Do you plan on exposing these to external apps, if yes, then how?

**Reply by twaik** [reply]
https://github.com/libffi/libffi . It is the most problematic because it has to be built using autoconf+configure+make. We can use a version built in libffi package (in `termux-packages`).

https://gitlab.freedesktop.org/wayland/wayland . It is less problematic. Android SDK does not support code generation and that is the reason I want to write scanner part in Java (to use it as Gradle task). The rest is simple C files that can be built with Android NDK makefiles.

**Reply by twaik** [reply]
libwayland also requires libexpat but I do not think it is needed (because scanner part of libwayland will not be used).

**Reply by agnostic-apollo** [reply]
Basically, if I am not wrong, you want https://github.com/termux/x11-packages/blob/master/packages/libwayland/build.sh and https://github.com/termux/termux-packages/tree/master/packages/libffi to be included in bootstrap? Or something else? And what else? What are deb sizes of the 4 archs of each package?

**Reply by twaik** [reply]
Not in bootstrap. In `termux-app` itself.

**Reply by twaik** [reply]
I need to use libwayland as JNI libraries.
That is the problem.

**Reply by agnostic-apollo** [reply]
I see, well currently I don't have much idea what you need included, create a POC and open a pull request so that I know what will need to get included and its safety and any performance affects and apk size increases.

**Reply by agnostic-apollo** [reply]
And what are the use cases you currently want to use this for?

**Reply by twaik** [reply]
It will let us use Android APIs inside programs without using shell. I know it sounds simple. But I think you will change your mind when you see a POC.
I need some time to prepare it.

**Reply by agnostic-apollo** [reply]
lolz, I already got that, I meant specific use case.

Yeah, take your time.

**Reply by twaik** [reply]
Oh, OK. Thanks.

**Reply by twaik** [reply]
I think usecases will be the same as of existing termux-api, but inside C programs.

**Reply by SDRausty** [reply]
> will be integrated into `termux-app` in coming weeks

Can `device admin app` permission for Termux also be granted?  Is this possible?

**Reply by agnostic-apollo** [reply]
>Can device admin app permission for Termux also be granted? Is this possible?

That is planned too with some APIs, but will be done after `termux-am-socket` is integrated into `termux-app`.

**Reply by agnostic-apollo** [reply]
Only issue with that is that it will put script kiddies and amateurs at risk who are running random code they blindly downloaded on their device, include device wipe. For such users I guess following warning will be given.

> It is not UNIX’s job to stop you from shooting your foot. If you so choose to do so, then it is UNIX’s job to deliver Mr. Bullet to Mr Foot in the most efficient way it knows. - Terry Lambert

**Reply by SDRausty** [reply]
@agnostic-apollo thank you for sharing.

**Reply by SDRausty** [reply]
>It is not UNIX’s job to stop you

Sounds like a good solution to me.

**Reply by twaik** [reply]
Hi guys. Sorry for a late reply.  I had a lot of work and did not have much time for a project. I have a question.
Wayland's protocols work in async way. Is there an rpc which works in synchronous way? Like AIDL/Binder code in Java on Android? RPC which can return a value to the calling thread?

**Reply by twaik** [reply]
I can try to port Binder of android-5.1 and write some  wrappers for Android/Termux api's if it is neccesary.

**Reply by twaik** [reply]
@agnostic-apollo I think I found a solution to use real Java code (including Binder) in C/C++ apps without calling `app_process` binary.

About using Java in C/C++
`app_process` is a single c++ file compiled to a binary and linked to libandroid_runtime. It contains a class with `AndroidRuntime` based class and `main` function with some helpers. 
`AndroidRuntime` class lets us to create JavaVM and load some classes we need and then execute code. We do not need to use shared libraries to link native functions into Java classes because we have `registerNativeMethods` function in `AndroidRuntime` (or `JniEnv`'s `RegisterNatives`). 
This API is private. But we can implement some kind of bridge library that will let us to use it without segfaults. Something like [that](https://github.com/fuyufjh/GraphicBuffer). Shortly: this library dlopens `libgui`, dlsyms needed methods, allocates some space (more than necessary) and calls constructors available in the system. I am not sure what to do with virtual functions but I think it is possible to make them work.
We can simply execute some Java code using JNI methods or we can use `callMain` function in case we have `main` in java class.

About connecting to service's Binder:
`termux-x11`' client's loader connects to the `termux-x11` app using code derived from `termux-am-library`. It simply sends intent to the app with Binder for callbacks. Not something complicated.

About code.
This api will depend on both library and `termux-app` version and I think it makes sense to put all AIDL's, Java code, native code, etc. inside `termux-app`'s apk and put symlinks to `$TERMUX_PREFIX`. That will let us avoid version mismatch errors.

What do you think?

**Reply by twaik** [reply]
Looks like invoking Java code from C++ programs is possible. We can link to `nativehelper` (which dlopens art or dalvik`, load art/dalvik` with [`JniInvocationInit`](https://cs.android.com/android/platform/superproject/+/master:libnativehelper/include_platform/nativehelper/JniInvocation.h;drc=f839d7916cf01eb25dbe46cf13d56b96c85dcca7;l=56) and then use [`JNI_CreateJavaVm`](https://cs.android.com/android/platform/superproject/+/master:libnativehelper/include_jni/jni.h;drc=321aca7238cd58a60c03dc416593abfed55ce449;l=1102). This will invoke all stuff we need.

**Reply by twaik** [reply]
@agnostic-apollo ?

**Reply by agnostic-apollo** [reply]
That "may" work, would have to test it on older devices 5,6,7 specially. `JNI_CreateJavaVM` apparently wasn't exported on older NDKs but it is now. There may be runtime restrictions for app user to do such stuff, loading system libraries is prohibited in new Android versions. Otherwise, app can load dex/jar files with `DexClassLoader` to get modified classloader and load required classes and execute functions. Execution should be possible in native code as well, should be possible to pass `Context` as well. Not sure how one would handle parameters and return calls though.

https://calebfenton.github.io/2017/04/05/creating_java_vm_from_android_native_code/

https://calebfenton.github.io/2017/04/14/calling_jni_functions_with_java_object_arguments_from_the_command_line/

https://github.com/rednaga/native-shim

**Reply by twaik** [reply]
Wow. Did not see those links. But it looks like event if we load Java code it would not be possible to use Android APIs without initializing them. [Android registers them right after creating VM](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/jni/AndroidRuntime.cpp;drc=1c8eb0109f716c7949a8ec09c1ece986f45e5c31;l=1670) and then [starts binder thread](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/cmds/app_process/app_main.cpp;drc=ce3b3291fd44906ee7cfcdedefaa806359ee9f8b;l=81). So anyway we should use libandroid-runtime. Somehow.
I do not think porting it to `termux-packages` would be complicated. If I am not wrong it uses `String8` and `String16` classes which have the only storage field containing pointer to char array, it uses Vector which has 4 fields we can easily emulate with struct. We can emulate them or simply port old android's `libutils`. They should be compatible with modern versions.

But anyway if we port this it will be slow. Creating vm in every process using termux api will consume a lot of resources. Now I am trying to create minimal Wayland java bindings that will save resources in `termux-app`/`termux-x11` applications and will let us use Java code with wayland without writing any JNI code.

**Reply by agnostic-apollo** [reply]
Invoking jvms is definitely slow, so should be avoided if it can be. Portingmay require aosp source tree to be included and behaviour would be different on different android versions and even vendors.

Not sure which `String` usage you are referring to but android native `String` handling, specially for binder parcels is complicated and different on different android versions.

Better way is service binders if just for termux-x11. If random code, then dex way would probably work too, custom objects may have issues, but gson could be used. I actually plan to look into that in future for `service call` commands to call android service interfaces, but not anytime soon.
